bin/opfp/src/cmd/util.rs

```diff
@@ -276,4 +279,186 @@ pub struct SystemConfig {
     pub overhead: B256,
     pub scalar: B256,
     pub gas_limit: u64,
-}
\ No newline at end of file
+}
+
+#[derive(Clone, Debug, Default, PartialEq, Eq)]
+pub struct VersionedState {
+    pub version: u8,
+    pub single_threaded_fpvmstate: SingleThreadedFPVMState,
+}
+
+#[derive(Clone, Debug, Default, PartialEq, Eq)]
+pub struct SingleThreadedFPVMState {
+    pub memory: Memory,
+    pub preimage_key: B256,
+    // assuming 32-bit machine
+    pub perimage_offset: u32,
+    pub cpu: CpuScalars,
+    pub heap: u32,
+    pub exit_code: u8,
+    pub exited: bool,
+    pub step: u64,
+    pub registers: [u32; 32],
+    pub last_hint: Vec<u8>,
+}
+
+#[derive(Clone, Debug, Default, PartialEq, Eq)]
+pub struct Memory {
+    pub pages: HashMap<u32, [u8; 4096]>,
+}
+
+#[derive(Clone, Debug, Default, PartialEq, Eq)]
+pub struct CpuScalars {
+    pub pc: u32,
+    pub next_pc: u32,
+    pub lo: u32,
+    pub hi: u32,
+}
+
+trait Decodeable {
+    fn decode<T>(&mut self, cursor: &mut Cursor<T>) -> Result<()>
+    where
+        T: AsRef<[u8]>;
+}
+
+impl TryFrom<Vec<u8>> for VersionedState {
+    type Error = String;
+
+    fn try_from(buffer: Vec<u8>) -> Result<Self, Self::Error> {
+        let mut v = VersionedState::default();
+        let mut cursor = Cursor::new(buffer);
+        let result = v.decode(&mut cursor);
+        return match result {
+            Ok(_) => Ok(v),
+            Err(_) => Err("invalid versioned state encoding".to_string()),
```

Comment:
It could be useful to embed the original error value here for debugging purposes